### DISCUSSION TITLE
Is it possible to port Janet's VM to WASM?

### DISCUSSION THREAD

**Discussion by glyh** [janet-lang/janet#947] [Q&A] [answered]
If so, we can run Janet bytecodes on WASM, without the whole interpretor. 

If that's possible, would it be faster? If I'm trying to implement such thing, what source files should I look into?

If that's not possible, what should an implementation be like to work well with WASM? through LLVM?

Thanks a lot!

------

P.S. seems like there's not yet a formal spec for Janet. I hope there will be one soon.
P.S.S If there will be no spec in a short period, could you tell me how should I read the source files?

**Comment by sogaiu** [comment]
Don't know if you've seen this already, but FWIW there is [this issue](https://github.com/janet-lang/janet/issues/542) regarding an official language spec.

**Reply by glyh** [reply]
Thanks, I've seen this.

**Comment by sogaiu** [comment, marked as answer]
Regarding reading the source, others may have some better tips, but here are some things I found useful (I've been trying to read and understand the source on and off):

* [janet.h](https://github.com/janet-lang/janet/blob/master/src/include/janet.h) can kind of be used as an outline for the public API
* The [official docs](https://janet-lang.org/docs/index.html) have some prose descriptions of various parts including but not limited to:
  * Data Structures
  * Fibers
  * The Janet Abstract Machine
  * The Event Loop
  * C API

(I mention the docs because I've found that they sometimes help in interpreting the source.)

I wrote up [some steps](https://github.com/sogaiu/index-janet-source/blob/master/doc/indexing-c.adoc) on getting an LSP server for C working so that code navigation can be done smoothly.  Possibly it might be of some use.

On a related note, recently I was shown [an article](https://mitchellh.com/writing/contributing-to-complex-projects)  on the general topic of trying to understand an existing code base and I found that it had some good tips.

There is a janet-language/help gitter channel / matrix room as well (in addition to the discussions area here on github) so it might be that some specific questions could be addressed there.
